File: leetcode/greedy/daily670.java
package greedy;

public class daily670 {
    public static void main(String[] args) {
        Solution670 solution = new Solution670();

        // case1
        System.err.println(solution.maximumSwap(2736));
        // case2
        System.err.println(solution.maximumSwap(9973));
    }
}

class Solution670 {
    public int maximumSwap(int num) {
        char[] s = String.valueOf(num).toCharArray();
        int n = s.length;
        int[] d = new int[n];
        for (int i = 0; i < n; i++) {
            d[i] = i;
        }
        for (int i = n - 2; i >= 0; i--) {
            if (s[i] <= s[d[i + 1]]) {
                d[i] = d[i + 1];
            }
        }
        for (int i = 0; i < n; i++) {
            int j = d[i];
            if (s[i] < s[j]) {
                char t = s[i];
                s[i] = s[j];
                s[j] = t;
                break;
            }
        }

        return Integer.parseInt(String.valueOf(s));
    }
}

File: leetcode/greedy/daily670.dart
void main(List<String> args) {
  var solution = Solution();

  // case1
  print(solution.maximumSwap(2736));
  // case2
  print(solution.maximumSwap(9973));
}

class Solution {
  int maximumSwap(int num) {
    List<int> s = num.toString().split('').map((e) => int.parse(e)).toList();
    int n = s.length;
    List<int> d = List<int>.generate(n, (i) => i);
    for (var i = n - 2; i >= 0; i--) {
      if (s[i] <= s[d[i + 1]]) {
        d[i] = d[i + 1];
      }
    }
    for (var i = 0; i < n; i++) {
      int j = d[i];
      if (s[i] < s[j]) {
        int t = s[i];
        s[i] = s[j];
        s[j] = t;
        break;
      }
    }

    return int.parse(s.join());
  }
}
